[AArch64]Fix invalid use of ld1/st1 in stack alloc

Lukacma opened the pull request:
This patch fixes invalid usage of scalar+immediate variant of ld1/st1 instructions during stack allocation caused by [c4bac7f](https://github.com/llvm/llvm-project/commit/c4bac7f7dcd931a5e561604e95656a24c3d1c9d9). This commit used ld1/st1 even when stack offset was outside of immediate range for this instruction, producing invalid assembly.

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sme-vg-to-stack.ll:
```diff
@@ -360,7 +355,17 @@ define void @vg_unwind_with_sve_args(<vscale x 2 x i64> %x) #0 {
 ; CHECK-NEXT:    .cfi_escape 0x10, 0x4f, 0x0a, 0x11, 0x60, 0x22, 0x11, 0x40, 0x92, 0x2e, 0x00, 0x1e, 0x22 // $d15 @ cfa - 32 - 64 * VG
 ; CHECK-NEXT:    addvl sp, sp, #-1
 ; CHECK-NEXT:    .cfi_escape 0x0f, 0x0d, 0x8f, 0x00, 0x11, 0x20, 0x22, 0x11, 0x98, 0x01, 0x92, 0x2e, 0x00, 0x1e, 0x22 // sp + 32 + 152 * VG
+; CHECK-NEXT:    mov x9, #256 // =0x100
```
This doesn't look right. I would have expected these offsets to be scaled by `vscale` (using `addvl`), but instead it's loading directly from address `sp + 256`.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -3354,10 +3360,21 @@ bool AArch64FrameLowering::spillCalleeSavedRegisters(
+      else {
+        // When stack offset out of range for st1b scalar + imm variant,
+        // store offset in register for use in scalar + scalar variant.
+        Register ScratchReg = findScratchNonCalleeSaveRegister(&MBB);
```
For the paired spills/fills where the offset is insufficient, it's probably simpler to just use the regular (single vector) spill/fill (LDR/STR) instructions, rather than having to find a scratch register and fiddle with the offsets.

Lukacma replied:
That's true, but it wouldn't it be better from performance perspective to do it this way ? Even when I need to use `addvl + madd` to get the correct offset, this should still be faster than having to issue a second load/store, especially since this code appears in the area where there is a load of  loads/stores and thus memory access units will be overloaded. Or am I missing something here ?

sdesmalen-arm replied:
> wouldn't it be better from performance perspective to do it this way

On the face of it I would probably agree that it would be better for performance reasons to use multi-vector stores. But I would only expect it to be necessary for 2 loads/stores at most. In the tests however, I see this being necessary for more loads/stores, which is suspicious.

Looking into this a bit more closely, I notice that with current `main` (i.e. without this PR), the offsets of the spills and fills are already incorrect. For some example:
```
declare void @bar()
define void @foo(<vscale x 4 x i32> %arg) nounwind {
  call void @bar()
  ret void
}
```
When building this with `llc -mattr=+sve2p1 < /tmp/t2.ll -enable-misched=false -enable-post-misched=false`, I see:
```
addvl   sp, sp, #-18
str     p15, [sp, #4, mul vl]           // 2-byte Folded Spill
str     p14, [sp, #5, mul vl]           // 2-byte Folded Spill
...
str     p4, [sp, #15, mul vl]           // 2-byte Folded Spill
ptrue   pn8.b
st1b    { z22.b, z23.b }, pn8, [sp, #4, mul vl] // 32-byte Folded Spill
st1b    { z20.b, z21.b }, pn8, [sp, #8, mul vl] // 32-byte Folded Spill
...
st1b    { z8.b, z9.b }, pn8, [sp, #32, mul vl] // 32-byte Folded Spill
```
Storing to `sp + 32 * VL` is incorrect, because the `addvl sp, sp, #-18` only allocated space for 18 vectors.
Further more, the offsets of the multi-vector stores are scaled by 4, instead of 2.

This is something that needs fixing. I would instead have expected:
```
st1b    { z22.b, z23.b }, pn8, [sp, #2, mul vl] // 32-byte Folded Spill
st1b    { z20.b, z21.b }, pn8, [sp, #4, mul vl] // 32-byte Folded Spill
...
st1b    { z8.b, z9.b }, pn8, [sp, #16, mul vl] // 32-byte Folded Spill
```
in which case only z8 and z9 would need to be spilled using the regular STR instructions, which isn't too bad.

I'm not sure if there is any data available on whether finding an extra scratch register + addvl is worth it in practice, but in the absence of that data I'd probably favour an approach that is lower in complexity.  From a quick look, it seems like the place where it tries to create the paired Z registers could be amended with an extra condition that checks the `ScalableByteOffset` to avoid creating the paired register when the offset > `((14 + 2) * 16)`.

Lukacma replied:
I have fixed the offset issue and make it revert STR when outside of range.

Lukacma committed:
[AArch64]Fix invalid use of ld1/st1 in stack alloc

This patch fixes invalid usage of scalar+immediate variant of ld1/st1 instructions during stack allocation caused by c4bac7f. This commit used ld1/st1 even when stack offset was outside of immediate range for this instruction, producing invalid assembly.

Lukacma commented:
As future work it might be useful to improve upon this work and enable using ST1B/LD1B even when stack offset is odd

sdesmalen-arm reviewed (approved):
LGTM, with two minor nits. Thanks for making the changes!

I agree there's follow-up work to support the case where either the offset is not a multiple of 2 or the first register is odd, but at least the code seems correct now.

Lukacma commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -3355,8 +3360,8 @@ bool AArch64FrameLowering::spillCalleeSavedRegisters(
-          .addImm(RPI.Offset) // [sp, #offset*scale],
-                              // where factor*scale is implicit
+          .addImm(RPI.Offset / 2) // [sp, #offset*2*scale],
+                                  // where scale is implicit
```
Yeah I was debating this comment. It depends what does the "offset" part in the comment mean. If it refers to RPI.Offset, then current comment is accurate, if it refers to RPI.Offset/2 then this change is correct. I thought that because RPI.Offset has "offset" in the name, I would  make "offset" in the comment refer to it.

sdesmalen-arm replied:
I read 'offset' as meaning the actual encoded immediate `RPI.Offset / 2`, because (along with the SP reg that is added on the line above), that is the MachineInstr that is being built here. The semantics of the encoded address is then `[sp, #offset * 2 * vscale]`, where `2 * vscale` is implied by the instruction, which is what I think the comment tries to clarify.

What about replacing `#offset` by `#imm`, such that we end up with:
```
// [sp, #imm*2*vscale]
// where 2*vscale is implicit
```
?

Lukacma committed:
Update llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

Co-authored-by: Sander de Smalen <sander.desmalen@arm.com>

Lukacma committed:
Fix comments and broken testcase

Lukacma committed:
Merge branch 'main' into sve_stack_alloc_ld1st1

Lukacma committed:
Fix comments
